ArduPlane: use Location::AltFrame for guided_state target frame references enabling ABOVE_TERRAIN

timtuxworth opened the pull request:
This PR changes the c++ code in Plane Guided mode to use Location:AltFrame for internal references to target Frame. 

This only applies for AP_PLANE_OFFBOARD_GUIDED_SLEW_ENABLED - effectively guided mode commands received via MAVLink, which could also be from Lua using the command_int Lua interface.

A very welcome side effect of this change is that it enables plane MAVLink CMD_GUIDED_CHANGE_ALTITUDE to work for  terrain altitudes in addition to the existing GLOBAL/Absolute and RELATIVE/Home altitudes.

Extracted from #27909 as requested by @IamPete1 to fix guided_state.target_alt_frame being a uint8_t whereas it actually stores AltFrame values from Location::AltFrame.

Similar to changes proposed for Sub in #27767 

Mavlink alt frame values (probably) needed for altitude navigation:

```
   MAV_FRAME_GLOBAL=0, /* Global (WGS84) coordinate frame + altitude relative to mean sea level (MSL). | */
   MAV_FRAME_GLOBAL_RELATIVE_ALT=3, /* 
          Global (WGS84) coordinate frame + altitude relative to the home position.
         | */
   MAV_FRAME_GLOBAL_TERRAIN_ALT=10, /* Global (WGS84) coordinate frame with AGL altitude (altitude at ground level). | */
```

whereas the internal Location:AltFrame values are 

```
    enum class AltFrame {
        ABSOLUTE = 0,
        ABOVE_HOME = 1,
        ABOVE_ORIGIN = 2,
        ABOVE_TERRAIN = 3
    };
```

So as you can see ABSOLUTE is co-incidentally the same (zero) but none of the other values match. Even worse MAV_FRAME_GLOBAL_RELATIVE_ALT=3 == AltFrame::ABOVE_TERRAIN=3.

So the MAVLink values need to be translated by GCS_Mavlink.cpp to values that can be used by Location and all of the other internal code that uses Location::ALtFrame.

IamPete1 commented on ArduPlane/Plane.h:
```diff
@@ -563,9 +563,10 @@ class Plane : public AP_Vehicle {
         // altitude adjustments
         float target_alt = -1;   // don't default to zero here, as zero is a valid alt.
         uint32_t last_target_alt = 0;
+        Location::AltFrame last_target_alt_frame = Location::AltFrame::ABSOLUTE;
```
This is unused.

timtuxworth replied:
In this PR yes, but when it's used in my other PR this was extracted from.

IamPete1 replied:
> Should I move this back over there?

Yes please.

timtuxworth replied:
Turns out I do need to use it to fix the CI fail

IamPete1 reviewed (commented):
This has broken CI. It also changes the logged value from being the MAVLink enum to the AP one.

timtuxworth commented:
> This has broken CI. It also changes the logged value from being the MAVLink enum to the AP one.

I think that is a bug. Why would plane.guided_state internally use the MAVLink enum for the target_alt_frame associated with the internal target_alt value it sends to TECS?

timtuxworth commented on ArduPlane/mode_guided.cpp:
```diff
@@ -140,6 +140,9 @@ void ModeGuided::update_target_altitude()
         int32_t delta_amt_i = (int32_t)(100.0 * delta_amt_f); 
         Location temp {};
         temp.alt = plane.guided_state.last_target_alt + delta_amt_i; // ...to avoid floats here, 
+        if(plane.guided_state.target_alt_frame == Location::AltFrame::ABOVE_HOME) {
+            temp.relative_alt = true;
+        }
```
Which is why it never would have worked. MAVLink was passing in the (MAVLink enum) Frame value which was never being used except for logging and was the wrong value to drive AP internal altitude frames. This fixes that.

IamPete1 replied:
I think you just need to delete this new if and you will pass CI.

timtuxworth replied:
True - but that is not actually correct. I think the real fix is that ModeGuiided::update_target_location() which is using last_target_alt should also use last_target_alt_frame.

timtuxworth replied:
Now implemented - and passes CI

timtuxworth commented:
> This has broken CI. It also changes the logged value from being the MAVLink enum to the AP one.

Is there an easy way to find out which test failed?

timtuxworth commented:
> This has broken CI. It also changes the logged value from being the MAVLink enum to the AP one.

I'm wondering if the test is valid? I think that switching out of GUIDED to LOITER and back to GUIDED should reset the plane.guided_state.target_alt back to -1, so I'm not sure why the test is expecting the altitude to be 70 +/-3 - can you explain that to me?

IamPete1 commented:
> I'm wondering if the test is valid? I think that switching out of GUIDED to LOITER and back to GUIDED should reset the plane.guided_state.target_alt back to -1, so I'm not sure why the test is expecting the altitude to be 70 +/-3 - can you explain that to me?

It was circling at 70m in guided. Changing to loiter should maintain that 70m altitude. Then changing back to guided again should still maintain 70m altitude. Until a command is received it should stay at the altitude at which the mode was entered.

timtuxworth commented:
> > I'm wondering if the test is valid? I think that switching out of GUIDED to LOITER and back to GUIDED should reset the plane.guided_state.target_alt back to -1, so I'm not sure why the test is expecting the altitude to be 70 +/-3 - can you explain that to me?
> 
> It was circling at 70m in guided. Changing to loiter should maintain that 70m altitude. Then changing back to guided again should still maintain 70m altitude. Until a command is received it should stay at the altitude at which the mode was entered.

Fixed - see my change to mode_guided.cpp

IamPete1 commented:
> Fixed - see my new change to mode_guided.cpp

I don't understand why we need to fix something that wasn't broken? Maybe I'm missing some case where we get it wrong currently?

timtuxworth commented:
> > Fixed - see my new change to mode_guided.cpp
> 
> I don't understand why we need to fix something that wasn't broken? Maybe I'm missing some case where we get it wrong currently?

I think it was broken, which caused the CI fail. It was using last_target_alt without a frame. The way I see it, it was purely coincidence that the sequence of events in the CI test would pass this test.

timtuxworth commented:
> It also changes the logged value from being the MAVLink enum to the AP one.

How do you think I should deal with this? I would argue it's a bug that it had been logging the MAVLink enum since that's not what the code is using, so it was probably not very useful.

amilcarlucas commented:
Are there static asserts that the AP internal frame defenition enums have the same values as the mavlink external frame defenition enums?

timtuxworth commented:
> Are there static asserts that the AP internal frame defenition enums have the same values as the mavlink external frame defenition enums?

They don't have the same values. That's the fundamental problem. See the first post on this PR, I lay out the mismatch.

clydemcqueen commented:
`GCS_MAVLINK::mavlink_coordinate_frame_to_location_alt_frame` is handy, and used by Copter and Sub.

https://github.com/ArduPilot/ardupilot/blob/1439aebf94ae388a82f28199aa7dd333c009b98c/libraries/GCS_MAVLink/GCS_Common.cpp#L6880

timtuxworth commented:
> `GCS_MAVLINK::mavlink_coordinate_frame_to_location_alt_frame` is handy, and used by Copter and Sub.
> 
> https://github.com/ArduPilot/ardupilot/blob/1439aebf94ae388a82f28199aa7dd333c009b98c/libraries/GCS_MAVLink/GCS_Common.cpp#L6880

Oh that's awesome, I didn't know that existed!

peterbarker commented on Tools/autotest/arduplane.py:
```diff
@@ -5529,6 +5531,26 @@ def MAV_CMD_GUIDED_CHANGE_ALTITUDE(self):
+        )
+        self.wait_altitude(alt-3, alt+3, timeout=30, relative=False)
+        self.change_mode('LOITER')
+        self.delay_sim_time(5)
```
Shouldn't there be a `wait_altitude` in here?

timtuxworth replied:
I've added an altitude_source = "TERRAIN_REPORT.current_height" to wait_altitude. It's a big clunky but it does work.

peterbarker commented on Tools/autotest/arduplane.py:
```diff
@@ -5529,6 +5531,26 @@ def MAV_CMD_GUIDED_CHANGE_ALTITUDE(self):
+            timeout=30,
+            relative=False,
+        )
+        self.delay_sim_time(5)
```
Why's this one in here?  Seems to be redundant after the `wait_altitude`
```suggestion
```

peterbarker commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -929,35 +929,33 @@ MAV_RESULT GCS_MAVLINK_Plane::handle_command_int_guided_slew_commands(const mavl
+        if (mavlink_coordinate_frame_to_location_alt_frame((MAV_FRAME)packet.frame, new_target_alt_frame)) {
+
+             // the requested alt data might be relative (to home)
+            if (new_target_alt_frame == Location::AltFrame::ABOVE_HOME) {
```
This should definitely still be a `switch` statement.

We should be rejecting packets if the new frame isn't one of ABOVE_HOME or ABSOLUTE

timtuxworth replied:
Interestingly, using mavlink_coordinate_frame_to_location_alt_frame() probably makes it work with ABOVE_TERRAIN "for free". Hence the changes I'm trying to make to the autotest to test that.

peterbarker commented on ArduPlane/Plane.h:
```diff
@@ -563,9 +563,11 @@ class Plane : public AP_Vehicle {
         uint32_t target_alt_time_ms = 0;
-        uint8_t target_alt_frame = 0;
+        Location::AltFrame target_alt_frame = Location::AltFrame::ABSOLUTE;
+        uint8_t target_mav_frame = -1;
```
Just leave at zero?

timtuxworth replied:
because the type has changed.

timtuxworth commented:
Thanks for looking at this @peterbarker - I've just make a whole bunch of changes which cleans it up a lot. Thanks for your questions/suggestions they lead to a lot of improvements. I'm pretty sure I fixed a couple of bugs in the original code.
@IamPete1 it would be great if you could take another look. I've now added autotests for ABOVE_TERRAIN which works quite nicely.

IamPete1 commented:
This seems to now add support for a new alt frame? Wasn't that going to be in https://github.com/ArduPilot/ardupilot/pull/27909?

timtuxworth commented:
> This seems to now add support for a new alt frame? Wasn't that going to be in #27909?

It was, but after adding mavlink_coordinate_frame_to_location_alt_frame() there doesn't seem much point. This gives us ABOVE_TERRAIN and I'd actually have to add code to remove it again.

IamPete1 commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -967,10 +959,14 @@ MAV_RESULT GCS_MAVLINK_Plane::handle_command_int_guided_slew_commands(const mavl
+        if (plane.guided_state.last_target_alt_frame == plane.guided_state.target_alt_frame && plane.guided_state.target_alt < plane.current_loc.alt) {
             plane.guided_state.target_alt_accel *= -1.0f;
         }
+        //if ((MAV_FRAME)packet.frame == MAV_FRAME_GLOBAL_TERRAIN_ALT) {
+        //    return MAV_RESULT_ACCEPTED;
+        // }
```
Remove commented code.

timtuxworth replied:
Done

IamPete1 commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -929,35 +929,27 @@ MAV_RESULT GCS_MAVLINK_Plane::handle_command_int_guided_slew_commands(const mavl
+            return MAV_RESULT_DENIED;
+        }
+
+        // compare two floats as near-enough
+        const float new_target_alt = packet.z * 100;
+        if (is_equal(plane.guided_state.target_alt,new_target_alt) && plane.guided_state.target_alt_frame == new_target_alt_frame) {
+            // no need to process any new packet/s with the same ALT any further, if we are already doing it.
+            return MAV_RESULT_ACCEPTED;
```
I know this stuff is there already, but I would be tempted to just remove the early return. I'm not sure what its really getting us.

timtuxworth replied:
I don't think that's a good idea. Removing it would set last_target_alt and last_target_alt_frame even if they are the same values as the incoming values. Leaving this in means those values are tell us the last actually different target values.

timtuxworth replied:
Removed

IamPete1 commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -967,10 +959,14 @@ MAV_RESULT GCS_MAVLINK_Plane::handle_command_int_guided_slew_commands(const mavl
-         // assign an acceleration direction
-        if (plane.guided_state.target_alt < plane.current_loc.alt) {
+         // assign an acceleration direction - this only works if the current and target frame are the same
+         // this should not be happening here - it should be in ModeGuided::update_target_altitude()
```
I agree, this is such a big rework now you might as well just fix it there. It also means it works with any alt frame and you don't need `last_target_alt_frame` anymore.

IamPete1 commented on ArduPlane/mode_guided.cpp:
```diff
@@ -139,13 +139,17 @@ void ModeGuided::update_target_altitude()
         int32_t delta_amt_i = (int32_t)(100.0 * delta_amt_f); 
         Location temp {};
-        temp.alt = plane.guided_state.last_target_alt + delta_amt_i; // ...to avoid floats here, 
+        int32_t temp_alt = plane.guided_state.last_target_alt + delta_amt_i;
```
```suggestion
        int32_t temp_alt = constrain_int32(plane.guided_state.target_alt, plane.guided_state.last_target_alt - delta_amt_i,  plane.guided_state.last_target_alt + delta_amt_i);
```
You can use the constrain trick so we don't need to know the sign of the accel.

IamPete1 replied:
That removes the if and the else.

IamPete1 replied:
Infact, this whole thing is velocity limiting not accel limiting. The command calls for a accel limit. There is either a error on the MAVLink spec or in the code.

timtuxworth replied:
Would it not be better to compare the new target location to the **current location** rather than the last target to determine the velocity required

timtuxworth replied:
Thanks -done.

IamPete1 commented on ArduPlane/mode_guided.cpp:
```diff
@@ -139,13 +139,17 @@ void ModeGuided::update_target_altitude()
-            temp.alt = MAX(plane.guided_state.target_alt, temp.alt);
+            temp_alt = MAX(plane.guided_state.target_alt, temp_alt);
         }
+        temp.set_alt_cm(temp_alt, plane.guided_state.target_alt_frame); 
```
This does not work correctly because the location is not initialized so it is not valid in terrain alt frame.

IamPete1 replied:
you need to store the last target location, not just its alt.

timtuxworth replied:
Done

IamPete1 commented on ArduPlane/Log.cpp:
```diff
@@ -491,11 +493,12 @@ const struct LogStructure Plane::log_structure[] = {
+// @Field: AltF: target alt frame (MAVLink)
 // @Field: Hdg:  target heading
 // @Field: HdgA: target heading lim
+// @Field: AltL: target alt frame (Location)
```
I don't see the benefit here, were logging the same information twice?

timtuxworth replied:
I want to know what came in via MAVLink _and_ what the result was (AltFrame)

IamPete1 reviewed (commented):
Quite a bit of the existing code looks suspect to me.....

peterbarker commented on ArduPlane/Log.cpp:
```diff
@@ -142,11 +143,12 @@ void Plane::Log_Write_OFG_Guided()
         target_heading         : guided_state.target_heading,
-        target_heading_limit   : guided_state.target_heading_accel_limit
+        target_heading_limit   : guided_state.target_heading_accel_limit,
+        target_alt_frame       : static_cast<uint8_t>(guided_state.target_location.get_alt_frame()),
```
rearrangment of field order is a little unfortunate here

timtuxworth replied:
It's not a reordering in terms of the values that end up in the log. The 6th field used to be called target_alt_frame but it contained the MAVLINK frame, so now the 6th field still contains the mavlink frame, the same value at the same position in the log and still called AltF - but it's source is a different value target_mav_frame.

peterbarker commented on Tools/autotest/vehicle_test_suite.py:
```diff
@@ -7282,11 +7282,18 @@ def get_altitude(self, relative=False, timeout=30, altitude_source=None):
                 altitude_source = "GLOBAL_POSITION_INT.alt"
+        if altitude_source == "TERRAIN_REPORT.current_height":
+            terrain = self.assert_receive_message('TERRAIN_REPORT')
+            if terrain is None:
+                raise NotAchievedException("Did not get TERRAIN_REPORT message")
```
```suggestion
```

You've asserted you've received it :-)

timtuxworth replied:
Thanks

peterbarker commented on Tools/autotest/vehicle_test_suite.py:
```diff
@@ -7282,11 +7282,18 @@ def get_altitude(self, relative=False, timeout=30, altitude_source=None):
         divisor = 1000.0  # mm is pretty common in mavlink
         if altitude_source == "SIM_STATE.alt":
             divisor = 1.0
+
```
Bad whitespace change

peterbarker commented on Tools/autotest/arduplane.py:
```diff
@@ -2942,8 +2942,6 @@ def TerrainLoiter(self):
             gpi = self.assert_receive_message('GLOBAL_POSITION_INT')
             terrain = self.assert_receive_message('TERRAIN_REPORT')
             rel_alt = terrain.current_height
-            self.progress("%um above terrain (%um bove home)" %
-                          (rel_alt, gpi.relative_alt/1000.0))
```
Why is this being removed?

Too noisy?

timtuxworth replied:
I thought I had put it in myself. I was trying to remove debug code I had added. I'll put it back.

timtuxworth replied:
Reverted

peterbarker replied:
Note quite; we print the final value just before dying so you know what the bad values were

timtuxworth replied:
Got it this time I think.

peterbarker commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -929,48 +929,32 @@ MAV_RESULT GCS_MAVLINK_Plane::handle_command_int_guided_slew_commands(const mavl
-        plane.guided_state.target_alt_frame = packet.frame;
-        plane.guided_state.last_target_alt = plane.current_loc.alt; // FIXME: Reference frame is not corrected for here
+        const int32_t new_target_alt = packet.z * 100;
+        if (is_equal(plane.guided_state.target_location.alt,new_target_alt) && plane.guided_state.target_location.get_alt_frame() == new_target_alt_frame) {
+            // no need to process any new packet/s with the same ALT any further, if we are already doing it.
+            // This is quite important as last_target_alt and last_target_alt_frame should only be set if the new one is different
+            return MAV_RESULT_ACCEPTED;
+        }
```
```suggestion
```

timtuxworth replied:
gone

Ryanf55 commented on libraries/AP_Scripting/examples/plane_guided_terrain.lua:
```diff
@@ -0,0 +1,229 @@
+    return protected_wrapper()
+else
+    gcs:send_text(MAV_SEVERITY.NOTICE,string.format("%s: Must run on Plane", SCRIPT_NAME_SHORT))
+end
```
```suggestion
end

```

timtuxworth replied:
Thanks - fixed.

Ryanf55 commented on Tools/autotest/arduplane.py:
```diff
@@ -5533,6 +5531,7 @@ def RunMissionScript(self):
+        self.start_subtest("set home relative altitude")
         self.takeoff(30, relative=True)
         self.change_mode('GUIDED')
         for alt in 50, 70:
```
Can we also test negative MSL altitudes in death valley? You mentioned it didn't work before, so now we can prove it does work.

timtuxworth replied:
I haven't tested this, follow up?

IamPete1 commented on ArduPlane/mode.cpp:
```diff
@@ -58,9 +58,11 @@ bool Mode::enter()
     plane.guided_state.target_heading_type = GUIDED_HEADING_NONE;
     plane.guided_state.target_airspeed_cm = -1; // same as above, although an airspeed of -1 is rare on plane.
-    plane.guided_state.target_alt = -1; // same as above, although a target alt of -1 is rare on plane.
+    //plane.guided_state.target_alt = -1; // same as above, although a target alt of -1 is rare on plane.
+    //plane.guided_state.target_alt_frame = Location::AltFrame::ABSOLUTE;
```
Remove commented code.

timtuxworth replied:
Done

IamPete1 commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -929,48 +929,32 @@ MAV_RESULT GCS_MAVLINK_Plane::handle_command_int_guided_slew_commands(const mavl
+        // param3 contains the desired vertical velocity (not acceleration)
         if (is_zero(packet.param3)) {
             // the user wanted /maximum acceleration, pick a large value as close enough
             plane.guided_state.target_alt_accel = 1000.0;
```
rename to `target_alt_rate`

timtuxworth commented on ArduPlane/Log.cpp:
```diff
@@ -129,9 +129,10 @@ struct PACKED log_OFG_Guided {
     float target_alt;
     float target_alt_accel;
-    uint8_t target_alt_frame;
+    uint8_t target_mav_frame;   // received MavLink frame
```
Yes but it's a MAV_FRAME value as defined in the mavlink common.xml

timtuxworth commented on ArduPlane/mode_guided.cpp:
```diff
@@ -138,15 +139,18 @@ void ModeGuided::update_target_altitude()
+        if (plane.current_loc.initialised() && plane.guided_state.target_location.initialised() && 
+            plane.current_loc.get_alt_cm(plane.guided_state.target_location.get_alt_frame(), target_alt_previous)) {
+            // create a new interim target location that that takes current_location and moves delta_amt_i in the right direction
+            int32_t temp_alt = constrain_int32(plane.guided_state.target_location.alt, target_alt_previous - delta_amt_i,  target_alt_previous + delta_amt_i);
```
done

timtuxworth commented on ArduPlane/mode_guided.cpp:
```diff
@@ -129,24 +129,29 @@ void ModeGuided::set_radius_and_direction(const float radius, const bool directi
+            // create a new interim target location that that takes current_location and moves delta_amt_i in the right direction
+            int32_t temp_alt_cm = constrain_int32(plane.guided_state.target_location.alt, target_alt_previous_cm - delta_amt_i,  target_alt_previous_cm + delta_amt_i);
+            Location temp_location = plane.guided_state.target_location;
+            temp_location.set_alt_cm(temp_alt_cm, target_frame);
```
Needed to be set_alt_cm() to take the target_frame.

timtuxworth commented:
Tested on a real plane and found one bug in the c++ code and the lua example needed some major rework.

timtuxworth commented:
I think this is "ready" - what do you think @IamPete1 ?

timtuxworth commented:
Rebased

timtuxworth commented:
Only 56 bytes on plane!

```
Binary Name      Text [B]       Data [B]     BSS (B)      Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  -------------  -----------  -----------  ----------------------------  -------------------------
antennatracker   0 (0.0000%)    0 (0.0000%)  0 (0.0000%)  0 (0.0000%)                                      621548
arducopter       0 (0.0000%)    0 (0.0000%)  0 (0.0000%)  0 (0.0000%)                                      151208
ardusub          0 (0.0000%)    0 (0.0000%)  0 (0.0000%)  0 (0.0000%)                                      360236
blimp            0 (0.0000%)    0 (0.0000%)  0 (0.0000%)  0 (0.0000%)                                      599496
arduplane        56 (+0.0031%)  0 (0.0000%)  0 (0.0000%)  56 (+0.0031%)                                    153736
ardurover        0 (0.0000%)    0 (0.0000%)  0 (0.0000%)  0 (0.0000%)                                      302576
arducopter-heli  0 (0.0000%)    0 (0.0000%)  0 (0.0000%)  0 (0.0000%)                                      152168
```
